blog/content/post/2020-08-06-race-condition-debugging.md

```diff
@@ -0,0 +1,210 @@
+date: 2020-08-06
+archives: "2020"
+
+tags:
+  - race condition
+  - go
+  - socketio
+  - debug
+  - debugging
+  - golang
+  - race
+  - debug
+categories:
+  - Firmware
+  - IoT
+  - Miscellaneous
+  - OS Dev
+  - App Dev
+  - Security
+  - Manufacturing
+
+---
+
+
+One of the hardest and exhaustive problems in software engineering are race
+conditions. They are sneaky, cunning, and nasty, often changing form, disguised
+under many different stack traces. In this article, I'll present my naive
+approach to this problem, why it wasn't the best option, and what tool can help
+to solve the issue in projects written in `Go`.
+
+### The problem
+One day, I was presented with the error in the [fork of the socketio library](https://github.com/3mdeb/socketio).
+The `socketio` is the library implementing real-time communication between
+a browser and a server. In my version of the library, there was a race condition
+error that sporadically led to a library-panic.
+
+### Debugging it the hard way
+When the program panics, the `Go` interpreter prints the stack trace on the
+standard output. The stack trace was my starting point on the hunt for the bug.
+A colleague also gave me the reproduction guide and the test repository with a
+simple app using the library to separate the problem from a project.
+The bug seemed to be connected with the
+[pingLoop()](https://github.com/3mdeb/socketio/blob/5e3badd8466ad335885957abcfdd1cc7aa8fe554/engineio/server_conn.go#L356)
+and the
+[NextWriter()](https://github.com/3mdeb/socketio/blob/5e3badd8466ad335885957abcfdd1cc7aa8fe554/engineio/websocket/server.go#L57)
+functions. After some testing, I got three unique stack traces and had some
+foggy ideas about the root of the problem, but the solution was still under my
+horizon. The problem in my debugging method was that I could see only one stack
+trace at the time, but race conditions are caused by two or more threads
+competing about the access to the shared resource. This quickly led me to use
+print statements in many places scattered in the code and lots of logs to
+analyze. Gathered logs confirmed my assumptions. At some point, two threads were
+fighting for access to one variable. The library was detecting the concurrency
+and throwing an error. Sadly this method swiftly became similar to trying to
+brute-force the solution of the problem. I had one part of the problem in the
+stack trace, but couldn't find the other half.
+
+### The better way to approach
+At some point, a colleague hinted me to try `-race` option of `Go`.
+```
+export GORACE=history_size=7
+go run -race main.go
+```
+Enabling this option was the solution to my problems. It quickly presented me
+with much more information than I could hope for. When the `-race` flag is
+enabled, `Go` carefully watches the code for any variable access, that could
+result in race condition. The only issue is, it sometimes appears to not work
+correctly, reporting a problem.
+`[failed to restore the stack]`
```

Comment:
Because of that problem, we export the history_size. See: https://golang.org/doc/articles/race_detector.html#Options
 
Without that, the would not be getting the stack at all. You can elaborate a little bit more about that option.